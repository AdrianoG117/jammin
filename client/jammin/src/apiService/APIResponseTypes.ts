export interface Jam {
  _id?: string
  title: string
  date: string
  description: string
  city: string
  cityCords: {
    lat: number
    lng: number
  }
  location: string
  locCords: {
    lat: number
    lng: number
  }
  host?: string
  numOfParticipants: number
  languages: string
  pastEvent: boolean
  comingEvent: true
  messages: Message[]
  __v?: number
}

export interface Message {
  name: string
  message:string
}

export interface User {
  firstname: string,
  lastname: string,
  email: string,
  password: string,
  pastEvents: Jam[],
  comingEvents: Jam[],
  _id?: string
  __v?: number
}