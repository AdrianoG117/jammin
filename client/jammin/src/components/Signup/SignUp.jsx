import React from 'react';
import './signup.css';

function SignUp() {
  return (
    <div className="signup-container">
      <h1>Create your account</h1>
      <form className="signup-form">
        <input
          required
          id="firstname"
          type="text"
          placeholder="First name"
          name="firstname"
        />
        <input
          required
          id="lastname"
          type="text"
          placeholder="Last name"
          name="lastname"
        />
        <input
          required
          id="email"
          type="text"
          placeholder="Email"
          name="email"
        />
        <input
          required
          id="password"
          type="password"
          placeholder="Password"
          name="password"
        />
        <button>JOIN NOW</button>
      </form>
    </div>
  );
}

export default SignUp;

//firstname LastName Email Password RepeatPassword
