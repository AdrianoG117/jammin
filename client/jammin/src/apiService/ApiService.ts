import {Jam, Message, User} from "./APIResponseTypes";

const BASE_URL = "http://localhost:3001";



const postEvent = (event: Jam): Promise<Jam | void> => {
  return fetch(`${BASE_URL}/jams`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(event),
  })
    .then((res) => res.json())
    .then((data:Jam) => data )
    .catch((err) => console.log(err));
};

interface searchValue {
    city: string
  }

const getJams = (city:searchValue): Promise<Jam[] | void> =>{
  return fetch(`${BASE_URL}/searchjam`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(city),
  })
    .then((res) => res.json())
    .then((data: Jam[]) => data)
    .catch((err) => console.log(err));
};

const postMessage = (message: Message, id: string): Promise<Message[] | void> =>  {
  return fetch(`${BASE_URL}/jams/${id}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(message),
  })
    .then((res) => res.json())
    .then((data: Message[]) => data)
    .catch(err => console.log(err));
};

const getEvent = (id:string): Promise<Jam[] | void> => {
  return fetch(`${BASE_URL}/jams/${id}`)
    .then((res) => res.json())
    .then((data: Jam[]) => data)
    .catch((err) => console.log(err));
};

const addParticipant = (id:string): Promise<Jam | void> => {
  return fetch(`${BASE_URL}/addparticipant`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(id),
  })
    .then((res) => res.json())
    .then((data: Jam) => data)
    .catch((err) => console.log(err));
};

const removeParticipant = (id:string): Promise<Jam | void> => {
  return fetch(`${BASE_URL}/removeparticipant`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(id),
  })
    .then((res) => res.json())
    .then((data: Jam) => data)
    .catch((err) => console.log(err));
};

//service for user

 const login = (user: { email: string, password: string,}): Promise<User | void> => {
  return fetch(`${BASE_URL}/login`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(user),
  })
    .then((res) => res.json())
    .then((data: User) => {
      data.comingEvents.sort(function (a, b) {
        return a.date < b.date ? -1 : a.date > b.date ? 1 : 0;
      });
      return data;
    })
    .catch((err) => console.log(err));
};

const register = (user: User): Promise<User | void> => {
  return fetch(`${BASE_URL}/register`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(user),
  })
    .then((res) => res.json())
    .then((data: User) => {
      console.log(data);
      data.comingEvents.sort(function (a, b) {
        return a.date < b.date ? -1 : a.date > b.date ? 1 : 0;
      });
      return data;
    })
    .catch((err) => console.log(err));
};

const addJam = (body:{id:string, jamId:string}):void => {
  fetch(`${BASE_URL}/addjam`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  })
    .then((res) => res.json())
    .catch((err) => console.log(err));
};

const removeJam = (body:{id:string, jamId:string}): void => {
  fetch(`${BASE_URL}/removejam`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  })
    .then((res) => res.json())
    .catch((err) => console.log(err));
};

export default {removeJam, addJam, register, postMessage, postEvent, getJams, getEvent, addParticipant, removeParticipant, login};
