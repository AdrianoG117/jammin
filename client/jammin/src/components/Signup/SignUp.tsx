import React, { useState, SetStateAction } from "react";
import apiService from "../../apiService/ApiService";
import "./signup.css";
import { useHistory } from "react-router-dom";
import { User } from "../../apiService/APIResponseTypes";

interface IProps {
  setUserData: React.Dispatch<SetStateAction< User >>; 
  setIsSignedUp: React.Dispatch<SetStateAction< boolean >>;
  state: User;
  setState: React.Dispatch<SetStateAction< User >>
}


const initialState:User = {
  firstname: "",
  lastname: "",
  email: "",
  password: "",
  pastEvents: [],
  comingEvents: [],
};

const SignUp:React.FunctionComponent<IProps> = ({ setUserData, setIsSignedUp }:IProps) => {
  const [state, setState] = useState(initialState);

  const history = useHistory();

  // adds name property based on it's relevant property/key name, then adds the corresponding value
  const handleChange = (event: React.FormEvent<HTMLInputElement>) => {
    const { name, value } = event.currentTarget;
    setState((previous) => ({
      ...previous,
      [name]: value,
    }));
  }

  // on submit adds the inital state object to the db
  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const user = apiService.register(state)
    .then((user) => { 
      if (user) setUserData(user); setIsSignedUp(true);setState(initialState);})
    .finally(() => [0])
    history.push("/dashboard");
  }

  return (
    <div className="signup-container">
      <h1>Create your account</h1>
      <form className="signup-form" onSubmit={handleSubmit}>
        <input
          required
          id="firstname"
          type="text"
          placeholder="First name"
          name="firstname"
          value={state.firstname}
          onChange={handleChange}
        />
        <input
          required
          id="lastname"
          type="text"
          placeholder="Last name"
          name="lastname"
          value={state.lastname}
          onChange={handleChange}
        />
        <input
          required
          id="email"
          type="text"
          placeholder="Email"
          name="email"
          value={state.email}
          onChange={handleChange}
        />
        <input
          required
          id="password"
          type="password"
          placeholder="Password"
          name="password"
          value={state.password}
          onChange={handleChange}
        />
        <button type="submit" className="signup-btn">JOIN NOW</button>
      </form>
    </div>
  );
}

export default SignUp;
