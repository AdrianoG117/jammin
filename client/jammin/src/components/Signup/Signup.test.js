import { render, screen } from "@testing-library/react";
import SignUp from "./SignUp";
import userEvent from "@testing-library/user-event";
import apiService from "../../apiService/ApiService";
<<<<<<< HEAD
import { useHistory } from "react-router";
=======
>>>>>>> dev_TS_refactor

const mockProps = {
    firstname: "front test",
    lastname: "last test",
    email: "email test",
    password: "password test",
    pastEvents: [],
    comingEvents: [],
  };

describe("SignUp component", ()=> {
  
  test("should match the snapshot", () => {
    const {container} = render(<SignUp
                  setUserData={()=>{}}
                  setIsSignedUp={()=>{}} />);
    expect(container.firstChild).toMatchSnapshot();
  });
  
  test("Should render the headings", () => {
    render (<SignUp
                  setUserData={()=>{}}
                  setIsSignedUp={()=>{}} />);
    //title
    screen.getByText(/Create your account/);
    screen.getByPlaceholderText(/First name/);
    screen.getByPlaceholderText(/Last name/);
    screen.getByPlaceholderText(/Email/);
    screen.getByPlaceholderText(/Password/);
    screen.getByRole("button", { name: "JOIN NOW" });
  });

//    submit function creates the object for the db
  test("Submit function creates object user when all fields filled in", async () => {
    
    const setUserData = jest.fn();
    const spy = jest.spyOn(apiService, "register");
    

    render (<SignUp 
                  setUserData={setUserData}
                  setIsSignedUp={()=>{}}
     />);

    const firstNameInput = screen.getByPlaceholderText(/First name/);
    const lastNameInput = screen.getByPlaceholderText(/Last name/);
    const emailInput = screen.getByPlaceholderText(/Email/);
    const passwordInput = screen.getByPlaceholderText(/Password/);
    const submitBtn = screen.getByRole("button", { name: "JOIN NOW" });

        // console.log(submitBtn);

    userEvent.type(firstNameInput, "front test");
    userEvent.type(lastNameInput, "last test");
    userEvent.type(emailInput, "email test");
    userEvent.type(passwordInput, "password test");
    userEvent.click(submitBtn);

    expect(spy).toHaveBeenCalledWith(mockProps);

    });

});