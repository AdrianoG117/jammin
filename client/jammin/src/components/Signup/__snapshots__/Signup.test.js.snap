// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup component should match the snapshot 1`] = `
<div
  class="signup-container"
>
  <h1>
    Create your account
  </h1>
  <form
    class="signup-form"
  >
    <input
      id="firstname"
      name="firstname"
      placeholder="First name"
      required=""
      type="text"
      value=""
    />
    <input
      id="lastname"
      name="lastname"
      placeholder="Last name"
      required=""
      type="text"
      value=""
    />
    <input
      id="email"
      name="email"
      placeholder="Email"
      required=""
      type="text"
      value=""
    />
    <input
      id="password"
      name="password"
      placeholder="Password"
      required=""
      type="password"
      value=""
    />
    <button
      class="signup-btn"
    >
      JOIN NOW
    </button>
  </form>
</div>
`;
